README.md
# Gaines and Kuklinski (2011) Estimators for Hybrid Experiments #

**GK2011** is a package for providing implementations of the treatment effect estimators for hybrid (self-selection) experiments, as developed by Gaines and Kuklinski (2011).

These functions estimate local average treatment effects for unobserved population subgroups inclined and disinclined to be treated, as revealed by a three-condition (two-arm) "hybrid" experimental design. In the design, participants are randomly assigned to one of three conditions: 1) treatment (T), 2) control (C), or 3) self-selection (S) of treatment or control. The design enables the estimation of four treatment effects:

 1. The sample average treatment effect,
 2. The effect for those inclined to choose treatment,
 3. The effect for those disinclined to choose treatment (or, equivalent, to inclined to choose control), and
 4. The naive difference in the outcome between those who chose treatment rather than control.



## Code Examples ##

The only function is `estimate()`, which provides a simple interface for estimating all effects:


```r
library("GK2011")

# create fake data
set.seed(12345)
d <- 
data.frame(rand = c(rep(1, 200), rep(0, 100)),
           tr = c(rep(0, 100), rep(1, 100), rep(0, 37), rep(1, 63)),
           y = c(rnorm(100), rnorm(100) + 1, rnorm(37), rnorm(63) + 3))

# estimate effects
estimate(rand = d$rand, tr = d$tr, y = d$y)
```

```
##   Effect   Estimate        SE         t            p
## 1      t  0.8000359 0.2076146  3.853467 1.570643e-04
## 2    t_s  2.5374464 0.3327255  7.626245 9.735505e-13
## 3    t_n -2.1582576 0.5848236 -3.690442 2.889546e-04
## 4  naive  2.9034906 0.3065426  9.471736 1.554308e-15
```

The response structure is a data.frame containing the name of each effect, the effect estimate, and bootstrapped standard error.

### Gaines and Kuklinski's data

A small example dataset is included that contains treatment and outcome data from Gaines and Kuklinski's (2011) study. The following code replicate the main descriptive results from Table 2 of their paper:


```r
data(ajps)
pmean <- function(x) sprintf("%0.1f", mean(x))
cbind(
  # Democrats
  aggregate(cbind(therm.mccain, therm.obama) ~ tr, data = ajps[ajps$pid == 1, ], FUN = pmean)[, 1:3],
  n_dem = aggregate(therm.obama ~ tr, data = ajps[ajps$pid == 1, ], FUN = length)[, 2],
  # Republicans
  aggregate(cbind(therm.mccain, therm.obama) ~ tr, data = ajps[ajps$pid == -1, ], FUN = pmean)[, 2:3],
  n_rep = aggregate(therm.obama ~ tr, data = ajps[ajps$pid == -1, ], FUN = length)[, 2]
)
```

```
##   tr therm.mccain therm.obama n_dem therm.mccain therm.obama n_rep
## 1  1         27.1        80.9    53         79.2        22.4    56
## 2  2         30.6        76.7    65         81.3        28.0    54
## 3  3         27.9        77.9    47         76.0        24.8    34
## 4  4         25.6        82.0    68         81.4        24.6    63
```

The paper reports four main sets of experimental results, separating the experiment by outcome (feeling thermometers for Obama and McCain) and by respondents' party identification (Democrats and Republicans).


Here are effects for McCain among Democrats:


```r
with(ajps[ajps$pid == 1, ], {
  estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.mccain)
})
```

```
##   Effect    Estimate       SE          t         p
## 1      t  -3.5175617 4.735196 -0.7428544 0.4590587
## 2    t_s -10.0114566 9.744498 -1.0273958 0.3056205
## 3    t_n   0.9708657 7.138275  0.1360084 0.8919784
## 4  naive   2.2546934 4.826405  0.4671579 0.6412786
```

Here are effects for McCain among Republicans:


```r
with(ajps[ajps$pid == -1, ], {
  estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.mccain)
})
```

```
##   Effect   Estimate       SE          t         p
## 1      t -2.1177249 3.419383 -0.6193296 0.5369916
## 2    t_s -5.0217865 8.877852 -0.5656533 0.5724746
## 3    t_n -0.5504535 4.673361 -0.1177853 0.9063932
## 4  naive -5.4444444 3.845384 -1.4158389 0.1600589
```

Here are effects for Obama among Democrats:


```r
with(ajps[ajps$pid == 1, ], {
  estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.obama)
})
```

```
##   Effect  Estimate       SE          t         p
## 1      t  4.201451 4.458794  0.9422843 0.3479887
## 2    t_s  8.762684 9.338501  0.9383394 0.3493344
## 3    t_n  1.048835 6.703290  0.1564657 0.8758550
## 4  naive -4.040989 4.675241 -0.8643381 0.3892172
```

Here are effects for Obama among Republicans:


```r
with(ajps[ajps$pid == -1, ], {
  estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.obama)
})
```

```
##   Effect   Estimate        SE           t         p
## 1      t -5.5522487  5.085318 -1.09181929 0.2773200
## 2    t_s -9.4531590 13.110036 -0.72106281 0.4719932
## 3    t_n -3.4469955  7.068427 -0.48766092 0.6264927
## 4  naive  0.1774043  5.689454  0.03118125 0.9751897
```


## Installation ##

[![CRAN](http://www.r-pkg.org/badges/version/GK2011)](https://cran.r-project.org/package=GK2011)
[![Build Status](https://travis-ci.org/leeper/GK2011.png?branch=master)](https://travis-ci.org/leeper/GK2011)
[![codecov.io](http://codecov.io/github/leeper/GK2011/coverage.svg?branch=master)](http://codecov.io/github/leeper/GK2011?branch=master)

This package is not yet on CRAN. To install the latest development version you can pull directly from GitHub:

```R
if(!require("ghit")){
    install.packages("ghit")
}
ghit::install_github("leeper/GK2011")
```


## References ##

Brian J. Gaines and James H. Kuklinski, (2011), "Experimental Estimation of Heterogeneous Treatment Effects Related to Self-Selection," *American Journal of Political Science* 55(3): 724-736, doi:10.1111/j.1540-5907.2011.00518.x.


NEWS
# CHANGES TO PACKAGENAME 0.1.2 #

* Fixed an error in calculation of standard errors. (#1)

# CHANGES TO PACKAGENAME 0.1.1 #

* Initial release.


tests/testthat.R
library("testthat")
library("GK2011")

test_check("GK2011")


tests/testthat/tests.R
context("Gaines and Kuklinski examples")

test_that("Results match AJPS (2011)", {
    data(ajps)
    e <- with(ajps, estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.obama, iterations = 100))
    expect_true(is.data.frame(e), label = "data.frame returned")
    expect_true(identical(dim(e), c(4L, 5L)), label = "data.frame dims correct")
    expect_true(identical(names(e), c("Effect", "Estimate", "SE", "t", "p")), label = "data.frame names correct")
    expect_true(length(attributes(e)$means) == 5, label = "treatment group means returned")
    expect_true(length(attributes(e)$alpha) == 1, label = "alpha returned")
})


R/ajps.R
#' @title Gaines and Kuklinski (2011) AJPS data
#' @aliases ajps
#' @description Subset of data from Gaines and Kuklinski (2011)
#' @details This dataset contains a subset of variables, extracted from the dataset used by Gaines and Kuklinski (2011).
#' @format
#' \describe{
#'   \item{tr}{The treatment indicator, where 1=treatment, 2=control, 3=chose treatment, 4=chose control.}
#'   \item{therm.obama}{A \dQuote{feeling thermometer} toward John McCain.}
#'   \item{therm.mccain}{A \dQuote{feeling thermometer} toward Barack Obama.}
#'   \item{pid}{An indicator of party identification, where -1=Republican, 0=Independent, 1=Democrat.}
#' }
#' @source Brian J. Gaines and James H. Kuklinski, (2011), "Experimental Estimation of Heterogeneous Treatment Effects Related to Self-Selection," \emph{American Journal of Political Science} 55(3): 724-736.
#' @examples
#' data(ajps)
#' 
#' # replicate Gaines and Kuklinski (2011) Table 2
#' pmean <- function(x) sprintf("%0.1f", mean(x))
#' cbind(
#'   # Democrats
#'   aggregate(cbind(therm.mccain, therm.obama) ~ tr, 
#'             data = ajps[ajps$pid == 1, ], FUN = pmean)[, 1:3],
#'   n_dem = aggregate(therm.obama ~ tr, 
#'                     data = ajps[ajps$pid == 1, ], FUN = length)[, 2],
#'   # Republicans
#'   aggregate(cbind(therm.mccain, therm.obama) ~ tr, 
#'             data = ajps[ajps$pid == -1, ], FUN = pmean)[, 2:3],
#'   n_rep = aggregate(therm.obama ~ tr, 
#'                     data = ajps[ajps$pid == -1, ], FUN = length)[, 2]
#' )
#' 
#' # effects for McCain among Democrats
#' with(ajps[ajps$pid == 1, ], {
#'   estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.mccain)
#' })
#' # effects for McCain among Republicans
#' with(ajps[ajps$pid == -1, ], {
#'   estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.mccain)
#' })
#'
#' # effects for Obama among Democrats
#' with(ajps[ajps$pid == 1, ], {
#'   estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.obama)
#' })
#' # effects for Obama among Republicans
#' with(ajps[ajps$pid == -1, ], {
#'   estimate(rand = tr %in% 1:2, tr = tr %in% c(1,3), y = therm.obama)
#' })
#' @seealso \code{\link{estimate}}
"ajps"


inst/ajps/subset.R
# packages
library("foreign")
library("car")

# load data
dat <- read.dta("gaineskuklinski-ajps2011.dta")
dat <- data[!is.na(data$sept_pho), ] # remove respondents with missing treatment data

# treatment variable
dat$cond.rand <- recode(dat$sept_pho,"'assigned positive flyers'=1;'assigned negative flyers '=-1;'assigned no flyers'=0;'opportunity to choose flyers'=9")
dat$cond.rand <- recode(as.numeric(dat$cond.rand),"1=3;2=1;3=2;4=9") #converting to numeric recodes -1=1;0=2;1=3;9=4; this recodes to final condition numbers

# clean up variable labels
dat$cond.choice <- gsub(",'"," ",dat$v1478_a) 
dat$cond.choice <- recode(data$cond.choice,'"yes  i ll look at the positive flyers."=1;"yes  i ll look at the negative flyers."=-1;"no thanks  i ll skip the flyers."=0')

# Temporary Conditions
#1=Randomized Control
#2=Randomized Pro Flyer
#3=Randomized Con Flyer
#4=Choice Control
#5=Choice Pro Flyer
#6=Choice Con Flyer

dat$conditions <- ifelse(dat$cond.rand == 9, dat$cond.choice, dat$cond.rand)

dat$tr <- ifelse(dat$conditions == 1, 2,
                 ifelse(dat$conditions == 3, 1,
                        ifelse(dat$conditions == 4, 4,
                               ifelse(dat$conditions == 6, 3, NA_real_))))
# subset data to exclude "pro" conditions
dat <- dat[!is.na(dat$tr), ]
# Final Conditions
#1=Randomized Con Flyer
#2=Randomized Control
#3=Choice Con Flyer
#4=Choice Control

# partyID
#'strong democrat'=1;
#'not very strong democrat'=.67;
#'lean democrat'=.33;
#'independent'=0;
#'lean republican'=-.33;
#'not very strong republican'=-.67;
#'strong republican'=-1;
# recode 'not sure' and NAs to 0/independent
dat$partyid <- recode(as.numeric(dat$scap8),"1=1;2=.67;3=.33;4=0;5=-.33;6=-.67;7=-1;8=0;NA=0")

# outcomes
dat$therm.mccain <- (as.numeric(dat$v1479_a)-1)/100 #rescaled 0-1; conversion to numeric adds 1 to 0-100 scale; 112 missing obs (no DKs)
dat$therm.obama <- (as.numeric(dat$v1480_a)-1)/100 #rescaled 0-1; conversion to numeric adds 1 to 0-100 scale; 107 missing obs (no DKs)

# subset for sharing
tmp <- na.omit(dat[, c("tr", "therm.obama", "therm.mccain", "pid")])

# rescale to match reported results
ajps$therm.obama <- 100*ajps$therm.obama
ajps$therm.mccain <- 100*ajps$therm.mccain

pmean <- function(x) sprintf("%0.1f", mean(x))
cbind(
    # Democrats
    aggregate(cbind(therm.mccain, therm.obama) ~ tr, data = tmp[tmp$pid == 1, ], FUN = pmean)[, 1:3],
    n_dem = aggregate(therm.obama ~ tr, data = tmp[tmp$pid == 1, ], FUN = length)[, 2],
    # Republicans
    aggregate(cbind(therm.mccain, therm.obama) ~ tr, data = tmp[tmp$pid == -1, ], FUN = pmean)[, 2:3],
    n_rep = aggregate(therm.obama ~ tr, data = tmp[tmp$pid == -1, ], FUN = length)[, 2]
)

ajps <- tmp

save(ajps, file = "ajps.RData")


R/estimate.R
#' @title estimate
#' @aliases estimate
#' @description Estimators for Hybrid Experiments
#' @param rand An integer or logical vector specifying whether each observation is from the random (1) or self-selection (0) arm of the experiment.
#' @param tr An integer or logical vector specifying whether each observation was treated (1) or control (0), regardless of the arm of the experiment.
#' @param y A numeric vector specifying outcome values.
#' @param iterations An integer specifying the number of bootstrap iterations used to estimate standard errors.
#' @details The package provides R implementations of the treatment effect estimators for hybrid (self-selection) experiments, as developed by Gaines and Kuklinski (2011). These functions estimate local average treatment effects for unobserved population subgroups inclined and disinclined to be treated, as revealed by a three-condition (two-arm) experimental design. In the design, participants are randomly assigned to one of three conditions: 1) treatment (T), 2) control (C), or 3) self-selection (S) of treatment or control. The design enables the estimation of three treatment effects:
#' \enumerate{
#'   \item First, the sample average treatment effect is estimated from conditions (1) and (2) as:\\
#'           \eqn{\hat{t} = \bar{Y}_{T} - \bar{Y}_{C}}
#'   \item The effect for those inclined to choose treatment is given by:\\
#'           \eqn{\hat{t}_s = \frac{\bar{Y}_{S} - \bar{Y}_{C}}{\hat{\alpha}}}
#'         where \eqn{\hat{\alpha}} is the observed proportion of individuals in group S that choose T (rather than C).
#'   \item The effect for those disinclined to choose treatment (or, equivalently, inclined to choose control) is given by:\\
#'            \eqn{\hat{t}_n = \frac{\bar{Y}_{T} - \bar{Y}_{S}}{1-\hat{\alpha}}}
#' }
#' By definition, the sample average treatment effect is an average of the other two effects.
#' @references Brian J. Gaines and James H. Kuklinski, (2011), "Experimental Estimation of Heterogeneous Treatment Effects Related to Self-Selection," \emph{American Journal of Political Science} 55(3): 724-736.
#' @return A data.frame containing the following variables:
#' \itemize{
#' \item \code{Effect}, a character vector of effect names (\dQuote{t}, \dQuote{t_s}, \dQuote{t_n}, \dQuote{naive})
#' \item \code{Estimate}, a numeric vector of effect estimates
#' \item \code{SE}, a numeric vector of bootstrapped standard errors
#' \item \code{t}, a t-statistic for the effect
#' \item \code{p}, a two-tailed p-value
#' }
#' The return value will also carry an attribute \dQuote{alpha}, indicating the estimated proportion \eqn{\alpha}.
#' @author Thomas J. Leeper <thosjleeper@gmail.com>
#' @examples
#' # create fake data
#' set.seed(12345)
#' d <- 
#' data.frame(rand = c(rep(1, 200), rep(0, 100)),
#'            tr = c(rep(0, 100), rep(1, 100), rep(0, 37), rep(1, 63)),
#'            y = c(rnorm(100), rnorm(100) + 1, rnorm(37), rnorm(63) + 3))
#' 
#' # estimate effects
#' estimate(rand = d$rand, tr = d$tr, y = d$y)
#' @importFrom stats sd pt
#' @seealso \code{\link{ajps}}
#' @export
estimate <- function(rand, tr, y, iterations = 5e3L) {
    
    # define estimators
    ## sate
    sate <- function(t, c) {
        mean(t) - mean(c)
    }
    ## alpha
    alpha <- function(st, sc) {
        (length(st)/length(c(st, sc)))
    }
    ## t_s
    t_s <- function(s, c, alpha) {
        (mean(s) - mean(c))/alpha
    }
     
    ## t_n
    t_n <- function(t, s, alpha) {
        (mean(t) - mean(s))/(1-alpha)
    }
    
    ## naive
    naive <- function(st, sc) {
        mean(st) - mean(sc)
    }
    
    # perform estimation
    te <- sate(y[rand == 1 & tr == 1], y[rand == 1 & tr == 0])
    a <- alpha(y[rand == 0 & tr == 1], y[rand == 0 & tr == 0])
    tes <- t_s(y[rand == 0], y[rand == 1 & tr == 0], a)
    ten <- t_n(y[rand == 1 & tr == 1], y[rand == 0], a)
    nv <- naive(y[rand == 0 & tr == 1], y[rand == 0 & tr == 0])
    
    # bootstrap standard errors
    ses <- replicate(iterations, {
        s <- sample(1:length(rand), length(rand), TRUE)
        alphatmp <- alpha(y[rand[s] == 0 & tr[s] == 1], y[rand[s] == 0 & tr[s] == 0])
        c(
          # SATE
          sate(y[rand[s] == 1 & tr[s] == 1], y[rand[s] == 1 & tr[s] == 0]),
          # t_s
          t_s(y[rand[s] == 0], y[rand[s] == 1 & tr[s] == 0], alphatmp),
          # t_n
          t_n(y[rand[s] == 1 & tr[s] == 1], y[rand[s] == 0], alphatmp),
          # naive
          naive(y[rand[s] == 0 & tr[s] == 1], y[rand[s] == 0 & tr[s] == 0])
          )
    })
    
    # output
    out <- list(Effect = c("t", "t_s", "t_n", "naive"),
                Estimate = c(te, tes, ten, nv),
                SE = c(sd(ses[1,]), sd(ses[2,]), sd(ses[3,]), sd(ses[4,]) ))
    out[["t"]] <- out[["Estimate"]]/out[["SE"]]
    dfs <- c(length(y[rand == 1]),
             length(y[(rand == 0) | (rand == 1 & tr == 0)]),
             length(y[(rand == 0) | (rand == 1 & tr == 1)]),
             length(y[rand == 0]))
    out[["p"]] <- 2*pt(abs(out[["t"]]), df = dfs - 1, lower.tail = FALSE)
    structure(out,
              class = "data.frame",
              row.names = 1:4,
              means = c(y_t = mean(y[rand == 1 & tr == 1]), 
                        y_c = mean(y[rand == 1 & tr == 0]),
                        y_s = mean(y[rand == 0]),
                        y_st = mean(y[rand == 0 & tr == 1]),
                        y_sc = mean(y[rand == 0 & tr == 01])),
              alpha = a)
}
